tests/query_params_test.go

```diff
@@ -0,0 +1,218 @@
+		"Host: localhost:8080\r\n" +
+		"\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(req), func(req *http.Request) {
+		want := url.Values{
+			"vegetable": []string{"potatO", "Tomato"},
+		}
+		got := req.URL.Query()
+		if diff := cmp.Diff(want, got); diff != "" {
+			t.Errorf("req.URL.Query() = %v, want %v, diff (-want +got):\n%s", got, want, diff)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+}
+
+func TestQueryParametersSameKeyDifferentCasing(t *testing.T) {
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+
+		if len(queryParams) != 2 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 2)
+		}
+
+		if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
+			t.Errorf("Expected one value for query parameter vegetables and one for Vegetables, got %d and %d", len(queryParams["vegetable"]), len(queryParams["Vegetable"]))
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+}
+
+// TestQueryParametersValidUnicode ensures keys and values that contain non-ASCII characters are parsed correctly
+func TestQueryParametersValidUnicode(t *testing.T) {
+	const value = "ăȚâȘî"
+	req := []byte("GET /?vegetable=ăȚâȘî HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if valueReceived := req.URL.Query()["vegetable"][0]; valueReceived != value {
+			t.Errorf("queryParams values: got %s, want %s", valueReceived, value)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+
+	key := "ăȚâȘî"
+	req = []byte("GET /?" + key + "=vegetable HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if listLen := len(req.URL.Query()[key]); listLen != 1 {
+			t.Errorf("len(queryParamsKey): got %d, want 1 value for %s", listLen, key)
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+}
+
+// Tests whether passing invalid Unicode will result  in a 400 Bad Request response
+func TestQueryParametersInvalidUnicodes(t *testing.T) {
```

Comment:
Can we have it similar to the testcases above where you are testing both the keys and the values?

Reply:
Resolved in [dca0c50](url)